Math/NumberTheory/DirichletCharacters.hs

```diff
@@ -0,0 +1,589 @@
+newtype DirichletCharacter (n :: Nat) = Generated [DirichletFactor]
+
+-- | The group (Z/nZ)^* decomposes to a product (Z/2^k0 Z)^* x (Z/p1^k1 Z)^* x ... x (Z/pi^ki Z)^*
+-- where n = 2^k0 p1^k1 ... pi^ki, and the pj are odd primes, k0 possibly 0. Thus, a group
+-- homomorphism from (Z/nZ)^* is characterised by group homomorphisms from each of these factor
+-- groups. Furthermore, for odd p, we have (Z/p^k Z)^* isomorphic to Z / p^(k-1)*(p-1) Z, an
+-- additive group, where an isomorphism is specified by a choice of primitive root.
+-- Similarly, for k >= 2, (Z/2^k Z)^* is isomorphic to Z/2Z * (Z / 2^(k-2) Z) (and for k < 2
+-- it is trivial).  (See `lambda` for this isomorphism).
+-- Thus, to specify a Dirichlet character, it suffices to specify the value of generators
+-- of each of these cyclic groups, when primitive roots are given. This data is given by a
+-- DirichletFactor.
+-- We have the invariant that the factors must be given in strictly increasing order, and the
+-- generator is as given by `generator`, and are each non-trivial. These conditions are verified
+-- using `validChar`.
+data DirichletFactor = OddPrime { _getPrime :: Prime Natural
+                                , _getPower :: Word
+                                , _getGenerator :: Natural
+                                , _getValue :: RootOfUnity
+                                }
+                     | TwoPower { _getPower2 :: Int -- this ought to be Word, but many applications
+                                                    -- needed to use wordToInt, so Int is cleaner
+                                                    -- Required to be >= 2
+                                , _getFirstValue :: RootOfUnity
+                                , _getSecondValue :: RootOfUnity
+                                }
+                     | Two
+
+instance Eq (DirichletCharacter n) where
+  Generated a == Generated b = a == b
+
+instance Eq DirichletFactor where
+  TwoPower _ x1 x2 == TwoPower _ y1 y2 = x1 == y1 && x2 == y2
+  OddPrime _ _ _ x == OddPrime _ _ _ y = x == y
+  Two              == Two              = True
+  _ == _ = False
+
+-- | A representation of <https://en.wikipedia.org/wiki/Root_of_unity roots of unity>: complex
+-- numbers \(z\) for which there is \(n\) such that \(z^n=1\).
+newtype RootOfUnity =
+  RootOfUnity { -- | Every root of unity can be expressed as \(e^{2 \pi i q}\) for some
+                -- rational \(q\) satisfying \(0 \leq q < 1\), this function extracts it.
+                fromRootOfUnity :: Rational }
+  deriving (Eq)
+
+instance Show RootOfUnity where
+  show (RootOfUnity q)
+    | n == 0    = "1"
+    | d == 1    = "-1"
+    | n == 1    = "e^(πi/" ++ show d ++ ")"
+    | otherwise = "e^(" ++ show n ++ "πi/" ++ show d ++ ")"
+    where n = numerator (2*q)
+          d = denominator (2*q)
+
+-- | Given a rational \(q\), produce the root of unity \(e^{2 \pi i q}\).
+toRootOfUnity :: Integral a => Ratio a -> RootOfUnity
```

Comment:
I'd rather avoid excessively polymorphic functions. Each call of `toRootOfUnity (2 % 3)` raises `Defaulting the following constraints to type ‘Integer’` warning. 

If you cast an argument to `Rational` first thing and do not use type `a` in any way, then it is better to make `toRootOfUnity` monomorphic, accepting only `Rational`. Users can easily combine it with `toRational` themselves, when needed.

Reply:
Changed, thanks